modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+
+----
+toUser(123, "John")
+toUser("123", true)
+----
+
+Both calls don't throw an exception. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +
+An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a Number. +
+
+DataWeave also uses a global type inference algorithm to validate your code even if no type is specified as
+constraints.
+
+Note that the use of constraints are optional but this can be useful in big scripts or multiple scripts where the type system could help you avoid bugs in your DataWeave logic by running the type checking algorithm.
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as String, Boolean, etc. These values are atomic, they are not composed by other values. The list of simple types are:
+
+* String
+* Boolean
+* Number
+* Regex
+* Null
+* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+Null is a type of only one value, which means that `null` can not be assigned to the type String or any other type except for the type Null.
+
+For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.
+
+----
+repeat("a", 3)
+----
+
+This returns `&quot;aaa&quot;`, but:
+
+----
+repeat(null, 3)
+----
+
+Throws an error because the value `null` can not be assigned to the parameter that expects a value of type String.
+
+----
+repeat(Null, 3)
+----
+
+In this case we are assigning to the first parameter a value that is the type Null (the type of this value
+is `Type&lt;Null&gt;`), and returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerce the value to be of type String (`&quot;Null&quot;`).
+
+Some other functions may have an overloaded signature for the function (the function has multiples definitions for different parameters types) that let it be called with `null`.
+
+For example, the map function has the signature `map(Array&lt;T&gt;, (T, Number)-&gt;R): Array&lt;R&gt;` (receives an Array of elements of type `T` (this is a type parameter), a function, and returns an Array of elements of type `R`) but it is also defined as `map(Null, (Nothing, Nothing)-&gt;Any): Null`. +
```

Comment:
@TomoDC are we covering type parameters here?

Reply:
No, I chose to explain the map function because it is one of the most used functions and the signature results to have the type parameter :S. We could change to some other function without type params like isNumeric from the dw::core::String (there are a lot of functions there defined for null)

Reply:
Maybe that would be best, cause otherwise you are forced to talk about a subject you haven't explained yet.

Reply:
@TomoDC Could you confirm the function to change and text?

Reply:
`For example, the **isNumeric** function from dw::core::Strings has the signature **isNumeric(String): Boolean** (receives a string and returns if that string is numeric) but it is also defined as **isNumeric(Null): Boolean**.

So if you call the function isNumeric with a null (isNumeric(null)) this does not throw an exception.
At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used. The function returns false because of the definition of isNumeric for null as the argument.`

Reply:
Updated.